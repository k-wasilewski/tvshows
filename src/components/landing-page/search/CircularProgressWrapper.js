import React from "react";
import {CircularProgress} from "@material-ui/core";
import {styles} from "../../../styles/styles";

export default function CircularProgressWrapper() {
    const classes = styles();

    return (
        <div className={classes.progress}>
            <CircularProgress />
        </div>
    )
}