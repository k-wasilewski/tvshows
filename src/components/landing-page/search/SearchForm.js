import {ErrorMsg} from "../../misc/ErrorMsg";
import React, {useEffect, useRef} from "react";
import {styles} from "../../../styles/styles";
import Button from '@material-ui/core/Button';

export function SearchForm(props) {
    const classes = styles();
    const inputRef = useRef();

    useEffect(() => {
        props.passInputRefToParent(inputRef);
    }, []);

    return (
        <div className={classes.dashboard}>
            <ErrorMsg msg={props.msg} />
            <input type='text' onChange={props.onChange} className={classes.input} ref={inputRef} />
            <Button variant='outlined' size='small' onClick={props.onSubmit}>Szukaj</Button>
            <Button variant='outlined' size='small' onClick={props.doReset}>Resetuj wyniki</Button>
        </div>
    )
}