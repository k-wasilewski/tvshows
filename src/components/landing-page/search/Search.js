import React from 'react';
import axios from 'axios';
import { connect } from 'react-redux';
import { setResults, setQuery } from '../../../redux/actions';
import SearchForm from "./SearchForm";
import CircularProgressWrapper from "./CircularProgressWrapper";

export class Search extends React.Component {
    constructor(props) {
        super(props);
        this._isMounted = false;

        this.state = {
            input: '',
            errorMsg: '',
            inputRef: null,
            loading: false
        };

        this.inputOnChange = this.inputOnChange.bind(this);
        this.submitQuery = this.submitQuery.bind(this);
        this.getInputRefFromChild = this.getInputRefFromChild.bind(this);
    }

    inputOnChange(event) {
        this.setState({input: event.target.value});
    }

    componentDidMount() {
        this._isMounted = true;
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    submitQuery(event) {
        const query = this.state.input;
        this.setState({input: '', loading: true});
        this.props.setQuery(query);
        const inputRef = this.state.inputRef;
        inputRef.value='';

        axios.get(`http://api.tvmaze.com/search/shows?q=${query}`)
            .then(resp => {
            if (this._isMounted) {
                if (query==='') {
                    if (this._isMounted)
                        this.setState({
                            errorMsg: 'Pole nie może być puste'
                        });
                } else if (resp.data.length===0) {
                    if (this._isMounted)
                        this.setState({
                            errorMsg: 'Nic nie znaleziono'
                        });
                } else {
                    if (this._isMounted) {
                        this.props.setResults(resp.data);
                        this.setState({
                            errorMsg: ''
                        });
                    }
                }
                this.setState({loading: false});
            }
        }).catch(error => {
            if (this._isMounted)
                this.setState({
                errorMsg: 'Błąd serwera'
            });
        });

        event.preventDefault();
    }

    resetResults = (event) => {
        const inputNode = this.state.inputRef;
        inputNode.value = '';
        this.setState({input: ''});
        this.props.setResults([]);
        this.props.setQuery('');
        event.preventDefault();
    }

    getInputRefFromChild(inputRef) {
        this.setState({inputRef: inputRef.current});
    }

    render() {
        return (
            <React.Fragment>
                {(this.state.loading) ? <CircularProgressWrapper /> : null}
                <SearchForm msg={this.state.errorMsg} onSubmit={this.submitQuery}
                            onChange={this.inputOnChange} doReset={this.resetResults}
                            passInputRefToParent={this.getInputRefFromChild} />
            </React.Fragment>

        );
    };
};

const mapStateToProps = (state) => {
    return {
        results: state.setResultsReducer.results
    };
};

const mapDispatchToProps = {
    setResults,
    setQuery
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);
