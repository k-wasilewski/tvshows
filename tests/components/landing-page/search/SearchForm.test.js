import React from 'react';
import renderer from 'react-test-renderer';
import {SearchForm} from "../../../../src/components/landing-page/search/SearchForm";
import {configure, mount} from 'enzyme';
import Adapter from "enzyme-adapter-react-16";

describe("SearchForm rendering specification", () => {
    it('SearchForm is rendered', () => {
        const component = renderer.create(
            <SearchForm/>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("SearchForm functional specification", () => {
    it('passInputRefToParent() is called when componentDidMount', () => {
        configure({adapter: new Adapter()});

        const mockPassInputRefToParent = jest.fn();

        const component = mount(
            <SearchForm passInputRefToParent={mockPassInputRefToParent}/>
        );

        expect(mockPassInputRefToParent).toHaveBeenCalled();
        component.unmount();
    });
});