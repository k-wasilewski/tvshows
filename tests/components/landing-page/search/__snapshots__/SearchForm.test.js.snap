// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm rendering specification SearchForm is rendered 1`] = `
<form
  className="makeStyles-dashboard-10"
  id="searchForm"
>
  <span
    className="makeStyles-errorMsg-6"
  />
  <input
    className="makeStyles-input-13"
    id="searchFormInput"
    type="text"
  />
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
    disabled={false}
    id="searchFormSubmitBtn"
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiButton-label"
    >
      Szukaj
    </span>
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Resetuj wyniki
    </span>
  </button>
</form>
`;
